plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  getSpan,
+  SpanStatusCode,
+  Span,
+  context,
+  diag,
+  SpanKind,
+} from '@opentelemetry/api';
+import type * as mongodb from 'mongodb';
+import {
+  MongodbCommandType,
+  MongoInternalCommand,
+  MongoInternalTopology,
+  WireProtocolInternal,
+  MongoDbInstrumentationConfig,
+  CursorState,
+} from './types';
+import { VERSION } from './version';
+import {
+  DatabaseAttribute,
+  GeneralAttribute,
+} from '@opentelemetry/semantic-conventions';
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+
+/** mongodb instrumentation plugin for OpenTelemetry */
+export class MongoDBInstrumentation extends InstrumentationBase<
+  typeof mongodb
+> {
+  constructor(protected _config: MongoDbInstrumentationConfig = {}) {
+    super('@opentelemetry/instrumentation-mongodb', VERSION, _config);
+  }
+
+  init() {
+    const { patch, unpatch } = this._getPatches();
+    return [
+      new InstrumentationNodeModuleDefinition<typeof mongodb>(
+        'mongodb',
+        this._supportedVersions,
+        undefined,
+        undefined,
+        [
+          new InstrumentationNodeModuleFile<WireProtocolInternal>(
+            'mongodb/lib/core/wireprotocol/index.js',
+            this._supportedVersions,
+            patch,
+            unpatch
+          ),
+        ]
+      ),
+    ];
+  }
+
+  private get _supportedVersions() {
```

Comment:
There is a built-in protected `supportedVersions` array https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-instrumentation/src/types.ts#L53

Is there a reason you chose to do it manually in the `init`?

Reply:
What do you mean by manually ? I believe we give it to the constructor for every instrumentation

Reply:
bah i'm sorry i got my wires crossed

Reply:
in fact the `supportedVersions` might be wrong if instrumentation patches multiple packages for example pg and pg-pool, http and https, so maybe we should remove this optional param anyway and then add version only when you create a new `InstrumentationNodeModuleDefinition` the same you do with name "mongodb" ?

Reply:
i'm fine with that but it might be confusing for the average instrumentation author who isn't a JS maintainer and is looking for the simplest way to patch a single package

Reply:
instead of
```
new InstrumentationNodeModuleDefinition<typeof mongodb>(
        'mongodb',
        this._supportedVersions,
        undefined,
```
do
```
new InstrumentationNodeModuleDefinition<typeof mongodb>(
        'mongodb',
        ['>=3.3 <4'],
        undefined,

```
I think that is simple enough ?

Reply:
@obecny I agree that we could remove it from the instrumentation interface.

> I think that is simple enough ?

I did that in the first place (and it worked) but i needed it twice so @Flarna suggested to have a single variable to avoid duplication.

Reply:
yea for version we can just declare some const that not necessary sits on class directly, but the main point is that it doesnt need to be declared as `_supportedVersions`. And because it was declared as optional we could remove it easily right ?

Reply:
We can remove anything at 0.x 🤣 

I think it is a relatively painless change for what we have so far.

Reply:
I haven't really understood why this fails but to me it looks like `InstrumentationNodeModuleDefinition` should copy the array instead of storing a reference.

I tried to reproduce this locally but failed. plugin-mongodb tests run fine so it doesn't seem to be a general setup issue on my machine.

Reply:
After re-reading this it might comes from the fact that i was using a private property even though a readonly is exposed on the abstract class ? I will try tomorrow but anyway that should not behave and "unset" the property

Reply:
It's not caused by the name. The problem is the sequence of calling `super` in constructor and initializing the properties of the derived class. `InstrumentationBase` calls `this.init()`.
See the generated JS code then it's clear why it's `undefined`:
```js
constructor(_config = {}) {
  super('@opentelemetry/instrumentation-mongodb', version_1.VERSION, _config);
  this._config = _config;
  this._supportedVersionsX = ['>=3.3 <4'];
}
```
It's similar as in C++ - never call virtual functions in a constructor...
In contrast to C++ JS calls the correct virtual function but on an object which is not completely constructed.

Reply:
We should add windows to the testing matrix

Reply:
> It's not caused by the name. The problem is the sequence of calling super in constructor and initializing the properties of the derived class. InstrumentationBase calls this.init().

Right i see, i will fix this then

Reply:
created https://github.com/open-telemetry/opentelemetry-js/issues/1989 as this seems to be an issue in instrumentation module.